### PR TITLE
Finish support for derived types (closes #363)

### LINKED ISSUE #363: Represent Fortran derived types in the PSyIR
Some sections of NEMO (notably those dealing with sea ice and ice bergs) make use of Fortran derived types. These are currently not handled by the PSyIR and are left as CodeBlocks which then means we can't have Kernels that contain derived types.

### PR DESCRIPTION
**Pull request by arporter**
A small PR that mainly updates the gocean1p0 implementation to use the new derived-type functionality where we previously had TODOs.

### CODE REVIEW

**Comment by arporter** [comment]
This PR removes the remaining TODOs that refer to #363. In several cases this really means replacing them with TODOs for #1028 since dependency analysis is not yet supported. There's also one for #935 because I need to change a `Symbol` into a `DataSymbol`.

File: src/psyclone/tests/domain/gocean/goloop_test.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    assert hsym.datatype == INTEGER_TYPE
+    assert isinstance(href, Reference)
+    # A derived-type reference must be in the form of a format string with
+    # "{0}" at the start.
+    with pytest.raises(NotImplementedError) as err:
+        loop._grid_property_reference("wrong%one")
+    assert ("Supplied grid property is a derived-type reference but does "
+            "not begin with '{0}': 'wrong%one'" in str(err.value))
+    gref = loop._grid_property_reference("{0}%grid%xstart")
+    assert isinstance(gref, Reference)
```
Comment (sergisiso) [inline src/psyclone/tests/domain/gocean/goloop_test.py:163]: isinstance(gref, StructureReference)

and maybe also check that gref.parent point to loop?

**Review by sergisiso — CHANGES_REQUESTED** [review]
@arporter This PR completes the structure PSyIR support and will closes #363. No TODOs for this issue are left in the code. I only made some minor in-line comments.

File: examples/psyir/create_aggregate_types.py [hunk 2]
```diff
@@ -41,9 +41,6 @@
 
 >>> python create_aggregate_types.py
 
-TODO #363 Ultimately this will output a Fortran representation of the PSyIR but
-this is a work in progress.
```
[the remark this answers is not in the document]
Comment (arporter) [reply examples/psyir/create_aggregate_types.py:45]: Yes, good idea. I've also renamed it to `create_structure_types` as I think that is more consistent.

File: src/psyclone/gocean1p0.py [hunk 3]
```diff
@@ -729,6 +730,57 @@ def add_bounds(bound_info):
             {'outer': {'start': data[3], 'stop': data[4]},
              'inner': {'start': data[5], 'stop': data[6]}}
 
+    # -------------------------------------------------------------------------
+    def _grid_property_reference(self, grid_property):
+        '''
+        Create a PSyIR reference using the supplied grid-property
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/gocean1p0.py:736]: Done and oops.

File: src/psyclone/gocean1p0.py [hunk 4]
```diff
@@ -729,6 +730,57 @@ def add_bounds(bound_info):
...
+        information (which will have been read from the config file.
+
+        :param str grid_property: the property of the grid for which to \
+            create a reference. This is the format string read from the \
+            config file or just a simple name.
+
+        :returns: A reference to a symbol.
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/gocean1p0.py:743]: Done.

File: src/psyclone/gocean1p0.py [hunk 5]
```diff
@@ -729,6 +730,57 @@ def add_bounds(bound_info):
...
+
+        if members[0] != "{0}":
+            raise NotImplementedError(
+                "Supplied grid property is a derived-type reference but does "
+                "not begin with '{{0}}': '{0}'".format(grid_property))
+
+        fld_sym = self.scope.symbol_table.lookup(self.field_name)
+        if type(fld_sym) == Symbol:
+            # We now know that this must be a DataSymbol.
+            # TODO to be replaced with functionality from #935.
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/gocean1p0.py:766]: Yes - moved the TODO, extended it to mention 1010 and disabled the pylint warnings.

File: src/psyclone/psyir/nodes/member.py [hunk 6]
```diff
@@ -38,6 +38,7 @@
 
 from __future__ import absolute_import
 from psyclone.psyir.nodes.node import Node
+from psyclone.core.access_info import AccessType
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/nodes/member.py:41]: Oops.

File: src/psyclone/psyir/nodes/structure_member.py [hunk 7]
```diff
@@ -81,8 +81,7 @@ def create(member_name, inner_member):
 
     def __str__(self):
         result = super(StructureMember, self).__str__() + "\n"
-        if self._children[0]:
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/nodes/structure_member.py:84]: Oh yes. I've improved it a bit as the way it was before was a hangover from when I permitted the first child to be None.

File: src/psyclone/psyir/nodes/structure_reference.py [hunk 8]
```diff
@@ -47,6 +47,7 @@
 from psyclone.psyir.symbols import DataSymbol, TypeSymbol, StructureType, \
     DeferredType
 from psyclone.errors import InternalError
+from psyclone.core.access_info import AccessType
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/nodes/structure_reference.py:50]: Oops.

File: src/psyclone/tests/domain/gocean/goloop_test.py (hunk 1 above)
Comment (arporter) [reply src/psyclone/tests/domain/gocean/goloop_test.py:163]: Oh yes, done.

**Comment by arporter** [comment]
Updates done and all tests OK. Ready for another look.

File: examples/psyir/Makefile [hunk 9]
```diff
@@ -37,7 +37,7 @@ include ../common.mk
 
 transform:
 	${PYTHON} create.py
-	${PYTHON} create_aggregate_types.py
+	${PYTHON} create_structure_types.py
```
Comment (sergisiso) [inline examples/psyir/Makefile:40]: I agree this is a better name, but there are a couple of references to it to update:
```
doc/user_guide/examples.rst
461:``create_aggregate_types.py`` demonstrates the representation of

examples/psyir/README.md
63:> python create_aggregate_types.py
```

**Review by sergisiso — CHANGES_REQUESTED** [review]
@arporter All code looks good now, there is just one comment about the new example name. Test pass and there are no pylint or pycodestyles issues introduced.

I also looked at the UserGuide PSyIR section and it could benefit from adding some info about the new structures nodes:
- At the Kernel-layer (although this is a bad name now, I should rename it to language-level nodes in #1010) can you list the new Reference and Member subclasses.
- At the tree navigation section the automethod to the member property is broken.
- At the structure datatype section, can you add and example similar to the one at the end of the array datatype part. (maybe the same from the example script?)
- Maybe at the end of the Creating PSyIR Nodes section there should be an example of the creation of a complex structure accessor as it may not be immediately obvious. (also a copy from the example script?)

I think that's it. I want to refactor this whole page but this will better be done after #1010

File: examples/psyir/Makefile (hunk 9 above)
Comment (arporter) [reply examples/psyir/Makefile:40]: Oops. Done now.

**Comment by arporter** [comment]
Thanks for the suggestions for the user guide. I think I've covered them all. Apologies for missing the other references to `create_aggregate_types` - I thought I'd caught them all! I've brought the branch up-to-date with master. Ready for another look once the CI is green.

**Review by sergisiso — APPROVED** [review]
@arporter Thanks for the documentation. It is all good now.

**Pull request merged**